mkosi/gentoo.py

```diff
@@ -0,0 +1,549 @@
+    }
+
+    @staticmethod
+    def try_import_portage() -> Dict[str, str]:
+        NEED_PORTAGE_MSG = "You need portage(5) for Gentoo"
+        PORTAGE_INSTALL_INSTRUCTIONS = """\
+        # Following is known to work on most systemd-based systems:
+        sudo tee /usr/lib/sysusers.d/acct-user-portage.conf > /dev/null <<- EOF
+        # /usr/lib/sysusers.d/portage.conf
+        u portage - "Portage system user" /var/lib/portage/home -
+        EOF
+
+        sudo systemd-sysusers --no-pager
+
+        sudo install --owner=portage --group=portage --mode=0755 --directory /var/db/repos
+        sudo install --owner=portage --group=portage --mode=0755 --directory /etc/portage/repos.conf
+        sudo install --owner=portage --group=portage --mode=0755 --directory /var/cache/binpkgs
+
+        sudo tee /etc/portage/repos.conf/eselect-repo.conf > /dev/null <<- EOF
+        [gentoo]
+        location = /var/db/repos/gentoo
+        sync-type = git
+        sync-uri = https://anongit.gentoo.org/git/repo/gentoo.git
+        EOF
+
+        git clone https://anongit.gentoo.org/git/proj/portage.git --depth=1
+        cd portage
+        tee setup.cfg > /dev/null <<- EOF
+        [build_ext]
+        portage-ext-modules=true
+        EOF
+
+        python setup.py build_ext --inplace --portage-ext-modules
+
+        sudo python setup.py install
+
+        sudo ln -s --relative /var/db/repos/gentoo/profiles/default/linux/amd64/17.1/no-multilib/systemd /etc/portage/make.profile
+        """
+        try:
+            from portage.const import (  # type: ignore
+                CUSTOM_PROFILE_PATH,
+                EBUILD_SH_ENV_DIR,
+                PROFILE_PATH,
+                USER_CONFIG_PATH,
+            )
+        except ImportError as e:
+            MkosiPrinter.warn(NEED_PORTAGE_MSG)
+            MkosiPrinter.info(PORTAGE_INSTALL_INSTRUCTIONS)
+            raise MkosiException(e)
+
+        return dict(profile_path=PROFILE_PATH,
+                    custom_profile_path=CUSTOM_PROFILE_PATH,
+                    ebuild_sh_env_dir=EBUILD_SH_ENV_DIR,
+                    portage_cfg_dir=USER_CONFIG_PATH)
+
+    def __init__(
+        self,
+        args: CommandLineArguments,
+        root: Path,
+        do_run_build_script: bool,
+    ) -> None:
+
+        ret = self.try_import_portage()
+
+        from portage.package.ebuild.config import config  # type: ignore
+
+        self.portage_cfg = config(config_root=str(root), target_root=str(root),
+                                  sysroot=str(root), eprefix=None)
+
+        PORTAGE_MISCONFIGURED_MSG = "You have portage(5) installed but it's probably missing defaults, bailing out"
+        # we check for PORTDIR, but we could check for any other one
+        if self.portage_cfg['PORTDIR'] is None:
+            die(PORTAGE_MISCONFIGURED_MSG)
+
+        if "build-script" in ARG_DEBUG:
+            for k, v in self.portage_cfg.items():
+                print(f"{k} = {v}")
```

Comment:
Maybe add another print above this, so that one can grep for e.g. "portage config" in the logs, otherwise this could be a `pprint.pprint`.